### ISSUE TITLE
Statically initialized objects in Status class

### ISSUE THREAD

**Issue by xyzzyz** [protocolbuffers/protobuf#1404]
The Status class has some statically initialized members for convenience (Status::OK, Status::UNKNOWN, and Status::CANCELLED). It is a good practice to avoid these, so I created a pull request that I thought would transparently remove them (https://github.com/google/protobuf/pull/1367). It turned out not to be as compatible with existing code as I though, so I'd like to step back and reconsider the situation.

It seems to me that if we want to remove static initializers, and Status has some statically initialized members, there's really no way other than removing those members or replacing them with something that can be e.g. linker-initialized. In any case, it involves changing the API.

The question is: is protobuf project open to such change? If so, we could start brainstorming on what would be the best and most convenient solution.

**Comment by pkasting** [comment]
A few notes:

* The desire to remove static initializers is coming from Chromium, where we have a strict policy against such.  However, I believe static instances of non-POD types are also banned by the Google style guide, and thus a change to remove these would be something libprotobuf maintainers should find desirable.

* Here are some ways of removing these static initializers beyond the pull request linked above:

(1) Replace caller instances of "status == Status::OK" with something like "status.ok()".  This requires ensuring sufficient functions similar to "ok()" exist to replace all static objects.
(2) Replace caller instances of "status == Status::OK" with something like "status == Status(error::OK)".  This requires adding an explicit Status(error::Code code) constructor.
(3) Replace caller instances of "status == Status::OK" with something like "status == Status::OK()".  This requires changing the static objects into static methods which return the relevant Status objects.

These are ordered by what I'd personally perceive as decreasing desirability in terms of how "hacky" they feel.

**Comment by pkasting** [comment]
Oh, also, it would be nice to get some sort of indication what the best to go is soon, so we can proceed with doing this locally in Chromium in parallel to trying to upstream that solution here.  This will unblock landing a long-delayed upgrade to our copy of libprotobuf.

**Comment by xfxyjwf** [comment]
@xyzzyz Could you take a look at the implementation of Google-internal util/task/status.h and mimic its implementation in protobuf's fork? I would like to avoid any API difference between the protobuf version and the Google-internal version.

**Comment by xyzzyz** [comment]
The internal version is pretty hacky, in that it depends on a) compiler generating the same memory layout for two different classes with the same fields in the order (which I'm not sure is guaranteed by the standard) b) making sure one doesn't update fields in Status without also updating ones in Status::Rep (as it will break terribly in runtime).

I could write a pull-request that would mimic the internal solution, but I wouldn't bet that it won't break across different compilers or platforms.

**Comment by xyzzyz** [comment]
Peter, as for your proposition (1), method .ok() already exists, think of rather things like "return Status::OK" which are quite popular throughout the codebase, so fixing cases like in (1) will not suffice.

**Comment by pkasting** [comment]
Yeah, if you want "return Status::OK" you'd need ideas 2 or 3.

I consulted some other people about the method used by the internal code and it seems like while it is likely to work in practice, it probably violates the C++ strict aliasing rules and thus is technically undefined behavior.  (You can cast from a Rep* to a containing Status or Status::Pod, but you can't go the other way, or cast across from one of the latter to the other.)  For more info see http://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing .

Unfortunately, given how widely referenced this is internally, it is probably infeasible to change the internal implementation.  Given this, I would reluctantly suggest copying the internal implementation and crossing fingers.  Luckily, the Status::Pod type is never used except in the block that does the static initialization here, so there are no places in which an optimization that assumes strict aliasing rules would introduce incorrect code.  Given this, the chance of something bad actually happening seems fairly remote.

**Comment by gerben-s** [comment]
So it seems Status is not really used in protobuf, only by some util code. We should kick that util code out of our lib.

**Comment by gerben-s** [comment]
I would like to move json out of core proto and depend on absl for status.

**Comment by gerben-s** [comment]
@Josh this bug is related to the discussion around moving to absl::Status.

**Comment by haberman** [comment]
Ok, there are a couple issues in play here:

1. Internally these constants (OK, CANCELLED, etc) have already been removed from `Status`. So our Status object is already incompatible with `absl::Status`.
2. Ultimately we may not be able to migrate to `absl::Status` anyway, because we expose `Status` in our public API. At the moment we have decided not to expose ABSL types directly in our public API. So our decision for now is to keep our own copy/port of `Status` indefinitely.
3. That still leaves us with the question of how to deal with these global constructors. I wonder if constexpr could help us create linker-initialized globals here.
